db/db_impl/db_impl_open.cc

```diff
@@ -313,8 +313,11 @@ Status DBImpl::NewDB(std::vector<std::string>* new_filenames) {
       new_filenames->emplace_back(
           manifest.substr(manifest.find_last_of("/\\") + 1));
     }
-  } else {
-    fs_->DeleteFile(manifest, IOOptions(), nullptr);
+  }
+  if (!s.ok()) {
+    fs_->DeleteFile(manifest, IOOptions(), nullptr).PermitUncheckedError();
+    fs_->DeleteFile(CurrentFileName(dbname_), IOOptions(), nullptr)
```

Comment:
I'm not sure the behavior here. Before we do the renaming, we have xxx.dbtmp and CURRENT. If rename failed, we may keep the same status xxx.dbtmp and CURRENT or we only have the CURRENT (new one). If it is the previous case, we will have not the CURRENT file at all?

Reply:
This is `NewDB()`. If `NewDB()` failed, then the `DB::Open()` call fails too. I think it's OK to delete all generated files here.

Reply:
I see. For newly created DB, it is fine to delete all of them, which will not create any lose.

Reply:
Furthermore, if we do not delete them. A subsequent attempt to create new db may fail because some FS does not support overwriting an existing file.

Reply:
> Furthermore, if we do not delete them. A subsequent attempt to create new db may fail because some FS does not support overwriting an existing file.

I did not read the code or subsequent attempt. In that case, does the sequence number increases? If so, we aways create a Manifest file with a new name and so the xxxx.dbtmp.

Reply:
I am a little confused here after reading the code.  The SetCurrentFile already deletes the temporary file if the rename failed so I am not sure why that is needed here.  I can see where the Manifest could be left around if the SetCurrentFile failed, but am not sure why we need both DeleteFile calls here.

Reply:
Thinking more. I think original code of deleting only the MANIFEST here is correct, but the current one has issues. If we delete the MANIFEST, we should also make sure the CURRENT file is deleted in this case. Otherwise, a future call to `DB::Open()` will find the CURRENT file and think the DB exists. If the MANIFEST is deleted, then open will fail. The difficulty here is that we cannot make sure CURRENT is deleted if we already end up at this point.
1. If the rename succeeded on server side, but client returns error. Then we cannot rely on the client to delete the CURRENT file.
2. If rename failed on remote, then we know CURRENT does not exist, and we are fine.

The original motivation for deleting the MANIFEST is to avoid future `NewDB()` call overwriting an existing file. Now I think this should not be done in the clean-up-after-error phase. Instead, this should be done at the beginning of `NewDB()`. Assuming no data loss. If a previous call to `NewDB()` failed, then the caller may retry `DB::Open()`. There are several possibilities.
1. Both CURRENT and MANIFEST exist. `NewDB()` won't be called, and the db is empty. This is OK.
2. Only the MANIFEST exists (maybe there is a .tmp file that failed to have been deleted). `NewDB()` will be called. In this case, we should avoid creating the same MANIFEST and tmp file.

Reply:
> > Furthermore, if we do not delete them. A subsequent attempt to create new db may fail because some FS does not support overwriting an existing file.
> 
> I did not read the code or subsequent attempt. In that case, does the sequence number increases? If so, we aways create a Manifest file with a new name and so the xxxx.dbtmp.

In `DeleteUnreferencedSst()`, we will bump the next_file_number_ to be the current largest file number + 1